projects/Icosa-Gallery/services/icosa-gallery/module.nix

```
@@ -0,0 +1,131 @@
+
+    database = {
+      host = lib.mkOption {
+        type = lib.types.str;
+        default = "127.0.0.1";
+        description = "Database host address.";
+      };
+
+      username = lib.mkOption {
+        type = lib.types.str;
+        default = "icosa-gallery";
+        description = "Database user.";
+      };
+
+      name = lib.mkOption {
+        type = lib.types.str;
+        default = "icosa-gallery";
+        description = "Database name.";
+      };
+    };
+
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = lib.types.attrsOf lib.types.str;
+        options = {
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for Icosa Gallery, will be passed as environment variables.
+        See <https://github.com/icosa-foundation/icosa-gallery/blob/main/django/django_project/settings.py>.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable (
+    lib.mkMerge [
+      {
+        environment.systemPackages = [ cfg.package ];
+
+        networking.firewall.allowedTCPPorts = lib.optional cfg.openFirewall cfg.port;
+
+        # https://docs.gunicorn.org/en/stable/deploy.html#systemd
+        systemd.services.icosa-gallery = {
+          wantedBy = [ "multi-user.target" ];
+
+          environment = {
+            POSTGRES_HOST = cfg.database.host;
+            POSTGRES_DB = cfg.database.name;
+            POSTGRES_USER = cfg.database.username;
+            BASE_DIR = "/var/lib/icosa-gallery";
+          }
+          // cfg.settings;
+
+          serviceConfig = {
+            Type = "notify";
+            NotifyAccess = "main";
+            User = "icosa-gallery";
+            Group = "icosa-gallery";
+            DynamicUser = true;
+            StateDirectory = "icosa-gallery";
+            WorkingDirectory = "/var/lib/icosa-gallery";
+            ExecStart = "${cfg.package}/opt/icosa-gallery/gunicorn django_project.wsgi:application --bind ${cfg.host}:${toString cfg.port} --timeout 900";
+            ExecReload = "${lib.getExe' pkgs.coreutils "kill"} -s HUP $MAINPID";
+            KillMode = "mixed";
+            TimeoutStopSec = "5";
+            PrivateTmp = true;
+            ProtectSystem = "strict";
+          };
+
+          preStart =
+            let
+              icosa-gallery = lib.getExe cfg.package;
+            in
+            ''
+              ${icosa-gallery} migrate
+              ${icosa-gallery} collectstatic --noinput
+              ${icosa-gallery} run_huey &
+            '';
+        };
+      }
+      (lib.mkIf (cfg.enableLocalDB) {
+        systemd.services.icosa-gallery.after = [ "postgresql.service" ];
+
+        services.postgresql = {
+          enable = true;
+          ensureDatabases = [ cfg.database.name ];
+          ensureUsers = [
+            {
+              name = cfg.database.username;
+              ensureDBOwnership = true;
+            }
+          ];
+        };
```

We've discussed this yesterday in our [best practices meeting](https://github.com/ngi-nix/ngipkgs/issues/1879#issuecomment-3640772881) and the general consensus was leaning towards not making local databases configurable.

The reason is that by default, postgresql connects through unix sockets, which [don't require authentication](https://www.postgresql.org/docs/current/auth-peer.html) for local connections. For this to work, we need the database name and user name to be the same.

```suggestion
        services.postgresql = {
          enable = true;
          ensureDatabases = [ "icosa-gallery" ];
          ensureUsers = [
            {
              name = "icosa-gallery";
              ensureDBOwnership = true;
            }
          ];
        };
```